File: web/share/js/kvm/msd.js
/*****************************************************************************
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
*****************************************************************************/


"use strict";


import {ROOT_PREFIX} from "../vars.js";
import {tools, $} from "../tools.js";
import {wm} from "../wm.js";


export function Msd() {
	var self = this;

	/************************************************************************/

	var __state = null;
	var __http = null;

	var __init__ = function() {
		$("msd-led").title = "Unknown state";

		tools.selector.addOption($("msd-image-selector"), "\u2500 Not selected \u2500", "");
		$("msd-image-selector").onchange = __selectImage;

		tools.el.setOnClick($("msd-download-button"), __clickDownloadButton);
		tools.el.setOnClick($("msd-remove-button"), __clickRemoveButton);

		tools.radio.setOnClick("msd-mode-radio", () => __sendParam("cdrom", tools.radio.getValue("msd-mode-radio")));
		tools.el.setOnClick($("msd-rw-switch"), () => __sendParam("rw", $("msd-rw-switch").checked));

		tools.el.setOnClick($("msd-select-new-button"), __toggleSelectSub);
		$("msd-new-file").onchange = __selectNewFile;
		$("msd-new-url").oninput = __selectNewUrl;
		$("msd-new-part-selector").onchange = __selectNewFile;

		tools.el.setOnClick($("msd-upload-new-button"), __clickUploadNewButton);
		tools.el.setOnClick($("msd-abort-new-button"), __clickAbortNewButton);

		tools.el.setOnClick($("msd-connect-button"), () => __clickConnectButton(true));
		tools.el.setOnClick($("msd-disconnect-button"), () => __clickConnectButton(false));

		tools.el.setOnClick($("msd-reset-button"), __clickResetButton);
	};

	/************************************************************************/

	self.setState = function(state) {
		if (state) {
			if (!__state) {
				__state = {"storage": {}};
			}
			if (state.enabled !== undefined) {
				__state.enabled = state.enabled;
				tools.feature.setEnabled($("msd-dropdown"), __state.enabled);
			}
			if (__state.enabled !== undefined) {
				if (state.online !== undefined) {
					__state.online = state.online;
				}
				if (state.busy !== undefined) {
					__state.busy = state.busy;
				}
				if (state.drive) { // Null on offline, ignore
					__state.drive = state.drive;
				}
				if (state.storage) { // Null on offline, ignore
					if (state.storage.parts !== undefined) {
						__state.storage.parts = state.storage.parts;
						__updateParts(__state.storage.parts);
					}
					if (state.storage.uploading !== undefined) {
						__state.storage.uploading = state.storage.uploading;
						__updateUploading(__state.storage.uploading);
					}
					if (state.storage.downloading !== undefined) {
						__state.storage.downloading = state.storage.downloading;
					}
					if (state.storage.images !== undefined) {
						__state.storage.images = state.storage.images;
					}
				}
				if (state.drive || (state.storage && state.storage.images !== undefined)) {
					__updateImageSelector(__state.drive, __state.storage.images);
				}
			}
		} else {
			__state = null;
		}
		__refreshControls();
	};

	var __refreshControls = function() {
		__updateControls(__state && (__state.online !== undefined) ? __state : null);
	};

	var __updateControls = function(state) {
		let o = (state && state.online);
		let d = (state ? state.drive : null);
		let s = (state ? state.storage : null);
		let busy = !!(state && state.busy);

		tools.hidden.setVisible($("msd-message-offline"), (state && !state.online));
		tools.hidden.setVisible($("msd-message-image-broken"), (o && d.image && !d.image.complete && !s.uploading));
		tools.hidden.setVisible($("msd-message-too-big-for-dvd"), (o && d.cdrom && d.image && d.image.size >= 33957083136));
		tools.hidden.setVisible($("msd-message-out-of-storage"), (o && d.image && !d.image.in_storage));
		tools.hidden.setVisible($("msd-message-rw-enabled"), (o && d.rw));
		tools.hidden.setVisible($("msd-message-another-user-uploads"), (o && s.uploading && !__http));
		tools.hidden.setVisible($("msd-message-downloads"), (o && s.downloading));

		tools.el.setEnabled($("msd-image-selector"), (o && !d.connected && !busy));
		tools.el.setEnabled($("msd-download-button"), (o && d.image && !d.connected && !busy));
		tools.el.setEnabled($("msd-remove-button"), (o && d.image && d.image.removable && !d.connected && !busy));

		tools.radio.setEnabled("msd-mode-radio", (o && !d.connected && !busy));
		tools.radio.setValue("msd-mode-radio", `${Number(o && d.cdrom)}`);

		tools.el.setEnabled($("msd-rw-switch"), (o && !d.connected && !busy));
		$("msd-rw-switch").checked = (o && d.rw);

		tools.el.setEnabled($("msd-connect-button"), (o && d.image && !d.connected && !busy));
		tools.el.setEnabled($("msd-disconnect-button"), (o && d.connected && !busy));

		tools.el.setEnabled($("msd-select-new-button"), (o && !d.connected && !__http && !busy));
		tools.el.setEnabled($("msd-upload-new-button"),
			(o && !d.connected && (tools.input.getFile($("msd-new-file")) || $("msd-new-url").value.length > 0) && !busy));
		tools.el.setEnabled($("msd-abort-new-button"), (o && __http));

		tools.el.setEnabled($("msd-reset-button"), (state && state.enabled && !busy));

		tools.el.setEnabled($("msd-new-file"), (o && !d.connected && !__http && !busy));
		tools.el.setEnabled($("msd-new-url"), (o && !d.connected && !__http && !busy));
		tools.el.setEnabled($("msd-new-part-selector"), (o && !d.connected && !__http && !busy));

		if (o && s.uploading) {
			tools.hidden.setVisible($("msd-new-sub"), false);
			$("msd-new-file").value = "";
			$("msd-new-url").value = "";
		}
		tools.hidden.setVisible($("msd-uploading-sub"), (o && s.uploading));
		tools.hidden.setVisible($("msd-new-tips"), (o && s.uploading && __http));

		let led_cls = "led-gray";
		let msg = "Unavailable";
		if (o && d.connected) {
			led_cls = "led-green";
			msg = "Connected to Server";
		} else if (o && s.uploading) {
			led_cls = "led-yellow-rotating-fast";
			msg = "Uploading new image";
		} else if (o && s.downloading) {
			led_cls = "led-yellow-rotating-fast";
			msg = "Serving the image to download";
		} else if (o) { // Sic!
			msg = "Disconnected";
		}
		$("msd-led").className = led_cls;
		$("msd-status").innerText = $("msd-led").title = msg;
	};

	var __updateUploading = function(uploading) {
		$("msd-uploading-name").innerText = (uploading ? uploading.name : "");
		$("msd-uploading-size").innerText = (uploading ? tools.formatSize(uploading.size) : "");
		if (uploading) {
			tools.progress.setPercentOf($("msd-uploading-progress"), uploading.size, uploading.written);
		}
	};

	var __updateParts = function(parts) {
		let names = Object.keys(parts).sort();
		{
			let writable = names.filter(name => (name === "" || parts[name].writable));
			let writable_json = JSON.stringify(writable);
			let el = $("msd-new-part-selector");
			if (el.__writable_json !== writable_json) {
				let sel = (el.value || "");
				el.options.length = 0;
				for (let name of writable) {
					let title = (name || "\u2500 Internal \u2500");
					tools.selector.addOption(el, title, name, (name === sel));
				}
				tools.hidden.setVisible($("msd-new-part"), (writable.length > 1));
				el.__writable_json = writable_json;
			}
		}
		{
			let names_json = JSON.stringify(names);
			let el = $("msd-storages");
			if (el.__names_json !== names_json) {
				el.innerHTML = names.map(name => `
					<div class="text">
						<div id="__msd-storage-${tools.makeTextId(name)}-progress" class="progress">
							<span class="progress-value"></span>
						</div>
					</div>
				`).join("<hr>");
				el.__names_json = names_json;
			}
		}
		for (let name of names) {
			let part = parts[name];
			let title = (
				name === ""
				? `${names.length === 1 ? "Storage: %s" : "Internal storage: %s"}` // eslint-disable-line
				: `Storage [${name}${part.writable ? "]" : ", read-only]"}: %s` // eslint-disable-line
			);
			let id = `__msd-storage-${tools.makeTextId(name)}-progress`;
			tools.progress.setSizeOf($(id), title, part.size, part.free);
		}
	};

	var __updateImageSelector = function(drive, images) {
		let sel = "";
		let el = $("msd-image-selector");
		el.options.length = 1;
		for (let name of Object.keys(images).sort()) {
			tools.selector.addSeparator(el);
			tools.selector.addOption(el, name, name);
			tools.selector.addComment(el, __makeImageSelectorInfo(images[name]));
			if (drive.image && drive.image.name === name && drive.image.in_storage) {
				sel = name;
			}
		}
		if (drive.image && !drive.image.in_storage) {
			sel = ".__external__"; // Just some magic name
			tools.selector.addOption(el, drive.image.name, sel);
			tools.selector.addComment(el, __makeImageSelectorInfo(drive.image));
		}
		el.value = sel;
	};

	var __makeImageSelectorInfo = function(image) {
		let text = `\xA0\xA0\xA0\xA0\xA0\u2570 ${tools.formatSize(image.size)}`;
		if (!image.complete) {
			text += ", broken";
		}
		if (image.in_storage !== undefined && !image.in_storage) {
			text += ", out of storage";
		}
		let ts = new Date(image.mod_ts * 1000);
		ts = new Date(ts.getTime() - (ts.getTimezoneOffset() * 60000));
		ts = ts.toISOString().slice(0, -8).replaceAll("-", ".").replace("T", "-");
		return `${text} \u2500 ${ts}`;
	};

	var __selectImage = function() {
		tools.el.setEnabled($("msd-image-selector"), false);
		tools.el.setEnabled($("msd-download-button"), false);
		tools.el.setEnabled($("msd-remove-button"), false);
		__sendParam("image", $("msd-image-selector").value);
	};

	var __clickDownloadButton = function() {
		let e_image = encodeURIComponent($("msd-image-selector").value);
		tools.windowOpen(`api/msd/read?image=${e_image}`);
	};

	var __clickRemoveButton = function() {
		let name = $("msd-image-selector").value;
		wm.confirm("Are you sure you want to remove this image?", name).then(function(ok) {
			if (ok) {
				tools.httpPost("api/msd/remove", {"image": name}, function(http) {
					if (http.status !== 200) {
						wm.error("Can't remove image", http.responseText);
					}
				});
			}
		});
	};

	var __sendParam = function(name, value) {
		tools.httpPost("api/msd/set_params", {[name]: value}, function(http) {
			if (http.status !== 200) {
				wm.error("Can't configure Mass Storage", http.responseText);
			}
			__refreshControls();
		});
	};

	var __clickUploadNewButton = function() {
		let file = tools.input.getFile($("msd-new-file"));
		__http = new XMLHttpRequest();
		let e_prefix = encodeURIComponent($("msd-new-part-selector").value);
		if (file) {
			let e_image = encodeURIComponent(file.name);
			__http.open("POST", `${ROOT_PREFIX}api/msd/write?prefix=${e_prefix}&image=${e_image}&remove_incomplete=1`, true);
		} else {
			let e_url = encodeURIComponent($("msd-new-url").value);
			__http.open("POST", `${ROOT_PREFIX}api/msd/write_remote?prefix=${e_prefix}&url=${e_url}&remove_incomplete=1`, true);
		}
		__http.upload.timeout = 7 * 24 * 3600;
		__http.onreadystatechange = __uploadStateChange;
		__http.send(file);
		__refreshControls();
	};

	var __uploadStateChange = function() {
		if (__http.readyState !== 4) {
			return;
		}
		if (__http.status !== 200) {
			wm.error("Can't upload image", __http.responseText);
		} else if ($("msd-new-url").value.length > 0) {
			let html = "";
			let msg = "";
			try {
				let end = __http.responseText.lastIndexOf("\r\n");
				if (end < 0) {
					end = __http.responseText.length;
				}
				let begin = __http.responseText.lastIndexOf("\r\n", end - 2);
				if (begin < 0) {
					end = 0;
				}
				let result_str = __http.responseText.slice(begin, end);
				let result = JSON.parse(result_str);
				if (!result.ok) {
					html = "Can't upload image";
					msg = result_str;
				}
			} catch (ex) {
				html = "Can't parse upload result";
				msg = `${ex}`;
			}
			if (html.length > 0) {
				wm.error(html, msg);
			}
		}
		tools.hidden.setVisible($("msd-new-sub"), false);
		$("msd-new-file").value = "";
		$("msd-new-url").value = "";
		__http = null;
		__refreshControls();
	};

	var __clickAbortNewButton = function() {
		__http.onreadystatechange = null;
		__http.abort();
		__http = null;
		__refreshControls();
		tools.hidden.setVisible($("msd-new-sub"), true);
	};

	var __clickConnectButton = function(connected) {
		tools.httpPost("api/msd/set_connected", {"connected": connected}, function(http) {
			if (http.status !== 200) {
				wm.error("Can't switch Mass Storage", http.responseText);
			}
			__refreshControls();
		});
		__refreshControls();
		tools.el.setEnabled($(`msd-${connected ? "connect" : "disconnect"}-button`), false);
	};

	var __clickResetButton = function() {
		wm.confirm("Are you sure you want to reset Mass Storage?").then(function(ok) {
			if (ok) {
				tools.httpPost("api/msd/reset", null, function(http) {
					if (http.status !== 200) {
						wm.error("Mass Storage reset error", http.responseText);
					}
				});
			}
		});
	};

	var __toggleSelectSub = function() {
		let el_sub = $("msd-new-sub");
		let visible = tools.hidden.isVisible(el_sub);
		tools.hidden.setVisible(el_sub, !visible);
		if (visible) {
			$("msd-new-file").value = "";
			$("msd-new-url").value = "";
		}
		__refreshControls();
	};

	var __selectNewFile = function() {
		let el = $("msd-new-file");
		let file = tools.input.getFile(el);
		if (file) {
			$("msd-new-url").value = "";
			if (__state && __state.storage && __state.storage.parts) {
				let part = __state.storage.parts[$("msd-new-part-selector").value];
				if (part && (file.size > part.size)) {
					let e_size = tools.escape(tools.formatSize(part.size));
					wm.error(`The new image is too big for the Mass Storage partition.<br>Maximum: ${e_size}`);
					el.value = "";
				}
			}
		}
		__refreshControls();
	};

	var __selectNewUrl = function() {
		if ($("msd-new-url").value.length > 0) {
			$("msd-new-file").value = "";
		}
		__refreshControls();
	};

	__init__();
}

File: kvmd/apps/kvmd/api/msd.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2023  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


import lzma
import time

from typing import AsyncGenerator

import aiohttp
import zstandard

from aiohttp.web import Request
from aiohttp.web import Response
from aiohttp.web import StreamResponse

from ....logging import get_logger

from .... import aiotools
from .... import htclient

from ....htserver import exposed_http
from ....htserver import make_json_response
from ....htserver import make_json_exception
from ....htserver import start_streaming
from ....htserver import stream_json
from ....htserver import stream_json_exception

from ....plugins.msd import BaseMsd

from ....validators import check_string_in_list
from ....validators.basic import valid_bool
from ....validators.basic import valid_int_f0
from ....validators.basic import valid_float_f01
from ....validators.net import valid_url
from ....validators.kvm import valid_msd_image_name


# ======
class MsdApi:
    def __init__(self, msd: BaseMsd) -> None:
        self.__msd = msd

    # =====

    @exposed_http("GET", "/msd")
    async def __state_handler(self, _: Request) -> Response:
        return make_json_response(await self.__msd.get_state())

    @exposed_http("POST", "/msd/set_params")
    async def __set_params_handler(self, request: Request) -> Response:
        params = {
            key: validator(request.query.get(param))
            for (param, key, validator) in [
                ("image", "name", (lambda arg: str(arg).strip() and valid_msd_image_name(arg))),
                ("cdrom", "cdrom", valid_bool),
                ("rw", "rw", valid_bool),
            ]
            if request.query.get(param) is not None
        }
        await self.__msd.set_params(**params)  # type: ignore
        return make_json_response()

    @exposed_http("POST", "/msd/set_connected")
    async def __set_connected_handler(self, request: Request) -> Response:
        await self.__msd.set_connected(valid_bool(request.query.get("connected")))
        return make_json_response()

    # =====

    @exposed_http("GET", "/msd/read")
    async def __read_handler(self, request: Request) -> StreamResponse:
        name = valid_msd_image_name(request.query.get("image"))
        compressors = {
            "": ("", None),
            "none": ("", None),
            "lzma": (".xz", (lambda: lzma.LZMACompressor())),  # pylint: disable=unnecessary-lambda
            "zstd": (".zst", (lambda: zstandard.ZstdCompressor().compressobj())),  # pylint: disable=unnecessary-lambda
        }
        (suffix, make_compressor) = compressors[check_string_in_list(
            arg=request.query.get("compress", ""),
            name="Compression mode",
            variants=set(compressors),
        )]

        async with self.__msd.read_image(name) as reader:
            if make_compressor is None:
                src = reader.read_chunked()
                size = reader.get_total_size()

            else:
                async def compressed() -> AsyncGenerator[bytes, None]:
                    assert make_compressor is not None
                    compressor = make_compressor()  # pylint: disable=not-callable
                    limit = reader.get_chunk_size()
                    buf = b""
                    try:
                        async for chunk in reader.read_chunked():
                            buf += await aiotools.run_async(compressor.compress, chunk)
                            if len(buf) >= limit:
                                yield buf
                                buf = b""
                    finally:
                        # Закрыть в любом случае
                        buf += await aiotools.run_async(compressor.flush)
                    if len(buf) > 0:
                        yield buf

                src = compressed()
                size = -1

            response = await start_streaming(request, "application/octet-stream", size, name + suffix)
            async for chunk in src:
                await response.write(chunk)
            return response

    # =====

    @exposed_http("POST", "/msd/write")
    async def __write_handler(self, request: Request) -> Response:
        unsafe_prefix = request.query.get("prefix", "") + "/"
        name = valid_msd_image_name(unsafe_prefix + request.query.get("image", ""))
        size = valid_int_f0(request.content_length)
        remove_incomplete = self.__get_remove_incomplete(request)
        written = 0
        async with self.__msd.write_image(name, size, remove_incomplete) as writer:
            chunk_size = writer.get_chunk_size()
            while True:
                chunk = await request.content.read(chunk_size)
                if not chunk:
                    break
                written = await writer.write_chunk(chunk)
        return make_json_response(self.__make_write_info(name, size, written))

    @exposed_http("POST", "/msd/write_remote")
    async def __write_remote_handler(self, request: Request) -> (Response | StreamResponse):  # pylint: disable=too-many-locals
        unsafe_prefix = request.query.get("prefix", "") + "/"
        url = valid_url(request.query.get("url"))
        insecure = valid_bool(request.query.get("insecure", False))
        timeout = valid_float_f01(request.query.get("timeout", 10.0))
        remove_incomplete = self.__get_remove_incomplete(request)

        name = ""
        size = written = 0
        response: (StreamResponse | None) = None

        async def stream_write_info() -> None:
            assert response is not None
            await stream_json(response, self.__make_write_info(name, size, written))

        try:
            async with htclient.download(
                url=url,
                verify=(not insecure),
                timeout=timeout,
                read_timeout=(7 * 24 * 3600),
            ) as remote:

                name = str(request.query.get("image", "")).strip()
                if len(name) == 0:
                    name = htclient.get_filename(remote)
                name = valid_msd_image_name(unsafe_prefix + name)

                size = valid_int_f0(remote.content_length)

                get_logger(0).info("Downloading image %r as %r to MSD ...", url, name)
                async with self.__msd.write_image(name, size, remove_incomplete) as writer:
                    chunk_size = writer.get_chunk_size()
                    response = await start_streaming(request, "application/x-ndjson")
                    await stream_write_info()
                    last_report_ts = 0
                    async for chunk in remote.content.iter_chunked(chunk_size):
                        written = await writer.write_chunk(chunk)
                        now = int(time.monotonic())
                        if last_report_ts + 1 < now:
                            await stream_write_info()
                            last_report_ts = now

                await stream_write_info()
                return response

        except Exception as err:
            if response is not None:
                await stream_write_info()
                await stream_json_exception(response, err)
            elif isinstance(err, aiohttp.ClientError):
                return make_json_exception(err, 400)
            raise

    def __get_remove_incomplete(self, request: Request) -> (bool | None):
        flag: (str | None) = request.query.get("remove_incomplete")
        return (valid_bool(flag) if flag is not None else None)

    def __make_write_info(self, name: str, size: int, written: int) -> dict:
        return {"image": {"name": name, "size": size, "written": written}}

    # =====

    @exposed_http("POST", "/msd/remove")
    async def __remove_handler(self, request: Request) -> Response:
        await self.__msd.remove(valid_msd_image_name(request.query.get("image")))
        return make_json_response()

    @exposed_http("POST", "/msd/reset")
    async def __reset_handler(self, _: Request) -> Response:
        await self.__msd.reset()
        return make_json_response()
